specs/capella/beacon-chain.md

```diff
@@ -381,19 +280,54 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
     process_sync_aggregate(state, block.body.sync_aggregate)
 ```
 
+#### New `get_expected_withdrawals`
+
+```python
+def get_expected_withdrawals(state: BeaconState) -> Sequence[Withdrawal]:
+    epoch = get_current_epoch(state)
+    withdrawal_index = state.next_withdrawal_index
+    index = ValidatorIndex((state.last_withdrawal_validator_index + 1) % len(state.validators))
+    ret: List[Withdrawal] = []
+    probed = 0
+    while (len(ret) < MAX_WITHDRAWALS_PER_PAYLOAD) and (probed < len(state.validators)):
+        val = state.validators[index]
+        balance = state.balances[index]
+        if is_fully_withdrawable_validator(val, balance, epoch):
+            withdrawal = Withdrawal(
+                index=withdrawal_index,
+                validator_index=index,
+                address=ExecutionAddress(val.withdrawal_credentials[12:]),
+                amount=balance,
+            )
+            ret.append(withdrawal)
+            withdrawal_index = WithdrawalIndex(withdrawal_index + 1)
+        elif is_partially_withdrawable_validator(val, balance):
+            withdrawal = Withdrawal(
+                index=withdrawal_index,
+                validator_index=index,
+                address=ExecutionAddress(val.withdrawal_credentials[12:]),
+                amount=balance - MAX_EFFECTIVE_BALANCE,
```

Comment:
may be there can be some `margin` balance that can still be left i.e. have `MAX_EFFECTIVE_BALANCE` + `MARGIN` as the minimum balance to maintain, this would resolve the issues with effectie balance droppin as the staker might not want his/her effective balance to drop by a step function

Reply:
Yeah this is something that @mcdee requested. I'm not against it, but I am cautious cause any value may be too high in the future and will require a hard fork to change.

Reply:
Doesn't the hysteresis take care of this? The actual balance would have to drop to 31.75 ETH for the effective balance to drop to 31 ETH. To lose 0.25 ETH under normal circumstances (not in a leak) takes quite a while, and in a leak situation adding a little margin won't help much in any case.

Reply:
The margin serves another purpose IMO and is to prevent to makes small withrawals. Currently they are so small that they may forbid creating on-chain proofs cause gas may be higher than the withrawal